### ISSUE TITLE
Rewarding active validators; Punishing inactive validators;

### ISSUE THREAD

**Issue by markwylde** [keppel/lotion#73]
@keppel 

Sorry if this isn't the right place to discuss this, but I'm sure it's a question more "noobs" like myself have.

Tendermint is great because it allows us to achieve consensus without mining. However lets say I want to reward validators by a "currency".

So one way I can do this can be seen in my [lotion-example/index.js](https://github.com/markwylde/lotion-example/blob/master/app/index.js):

```js
app.useBlock((state, chainInfo) => {
  console.log({state, chainInfo})
  Object.keys(chainInfo.validators)
    .forEach(validator => {
      state.ledger.push({
        to: validator,
        from: '_network',
        at: chainInfo.height,
        amount: 0.0001
      })
    })
})
```

So this will, for every block, add a transaction of 0.0001. I'm aware that the transaction log itself could keep a track of this, meaning I could just have a key value pair for each address, but just for my learning I'm putting the transactions into the state itself. I guess I should be doing:
```js
app.useBlock((state, chainInfo) => {
  console.log({state, chainInfo})
  Object.keys(chainInfo.validators)
    .forEach(validator => {
      state.accounts[validator] = state.accounts[validator] + 0.0001 || 0
    })
})
```

So, the above works great. However it always adds 0.0001 for every validator for every block.

Ideally I only want to reward a validator if they are online and have validated the block.

Maybe something like:

```js
app.useBlock((state, chainInfo) => {
  console.log({state, chainInfo})
  Object.keys(chainInfo.validators)
    .forEach(validator => {
      if (chainInfo.activeValidator[validator]) {
        state.accounts[validator] = state.accounts[validator] + 0.0001 || 0
      } else {
        state.accounts[validator] = state.accounts[validator] - 0.001 || 0
      }
    })
})
```

tl;dr; I guess the brief question is: Is there any way to verify if a validator is active or not?

Thanks again

**Comment by keppel** [comment]
hey Mark! great question. there's currently no way to do this, definitely need to fix that.

I like the way you proposed above. block handlers should be able to see which validators signed the block being handled, then a strategy for keeping track of "active validators" can be implemented as middleware.

I'll leave this issue open until that's done. thanks Mark :)

**Comment by markwylde** [comment]
Hey @keppel 

So I should have updated. I found a (far from complete) solution. I've updated my repository but let me explain what I have done.

First of all I had to change the lotion codebase to use asynchronous middleware. As per your comment in issue #7 I have not made a PR. Also my change actually prevents the app from re validating the blockchain on restart, presumably because the Tendermint server has not started accepting REST requests. My change to the code base can be seen here:
https://github.com/markwylde/lotion/commit/253c101dff7367d94859fb225aea4d57a91b9d3d

So now that we have asynchronous  middleware, we can now call back to the tendermint api to get some information.

I found the blocks api, which allows me to query this:
```bash
http://localhost:46650/block?height=${chainInfo.height - 1}
```

This returns an object containing all the validators who responded for that request (I believe. Please correct me if I'm wrong).

As you can see from [line 24 in my example](https://github.com/markwylde/lotion-example/blob/master/app/index.js#L24):

```javascript
app.useBlock(async (state, chainInfo) => {
  if (chainInfo.height > 2) {
    const lastBlock = await axios(`http://localhost:46650/block?height=${chainInfo.height - 1}`)
    const validatorsMeta = (await axios(`http://localhost:46650/validators?height=${chainInfo.height - 1}`)).data.result.validators

    const activeValidators = Array.from(new Set(
      lastBlock.data.result.block.last_commit.precommits
        .filter(el => !!el)
        .map(validator => validator.validator_address)
    ))

    // Add a block for each validator
    let longestValidator
    activeValidators.forEach(validatorId => {
      validatorId = validatorId.toString()
      state.validators[validatorId] = state.validators[validatorId] || {
        lastWinningBlock: 0,
        missedBlocks: 0,
        amount: -1
      }

      state.validators[validatorId].lastWinningBlock++

      if (state.validators[validatorId].lastWinningBlock > (state.validators[longestValidator] || {lastWinningBlock:0}).lastWinningBlock) {
        longestValidator = validatorId
      }
    })

    state.validators[longestValidator].lastWinningBlock = 0
    state.validators[longestValidator].amount += 1

    // Add penalty for bad validators
    Object.keys(chainInfo.validators)
      .forEach(publicKey => {

        const validator = validatorsMeta.find(v => {
          return `01${v.pub_key.data.toLowerCase()}` === publicKey
        })
        const validatorAddress = validator.address

        if (!activeValidators.includes(validatorAddress)) {
          state.validators[validatorAddress].missedBlocks++
          if (state.validators[validatorAddress].missedBlocks > 4) {
            console.log(`Deleting validator "${publicKey}"`)
            console.log(chainInfo.validators[publicKey])
            delete chainInfo.validators[publicKey]
            delete state.validators[validatorAddress]
          }
        }
})
```

This appears to be working. The only issue is I can't call those endpoints before the server is running, therefore I can't verify the blockchain.

Anyway, have a look at my final result:
https://github.com/markwylde/lotion-example/blob/master/README.md)

I'd lovey our opinion, I'm I going in completely the wrong direction?

**Comment by keppel** [comment]
Mark, this looks great, you're definitely going in the right direction. Rather than supporting async middleware and talking to tendermint RPC to get data about which validators signed a block, I'd rather have the data that you're currently accessing in `lastBlock.data.result.block` passed in as an argument to block handlers.

**Comment by bradleat** [comment]
@keppel is there a lastBlock param avail without #133?
